[gRIBI-4.1] Add test for PARAMS_DIFFER_FROM_OTHER_CLIENTS

deep-gajjar5 opened the pull request:
Connect clientA with SINGLE_PRIMARY client redundancy. Connect clientB specifying ALL_PRIMARY client redundancy, ensuring that the connection is rejected with `PARAMS_DIFFER_FROM_OTHER_CLIENTS`.

robshakir commented on compliance/election.go:
```diff
@@ -132,6 +132,49 @@ func TestDifferingElectionParameters(c *fluent.GRIBIClient, t testing.TB, opts .
 	)
 }
 
+// TestParamsDifferFromOtherClients checks that when a client A is connected with election ID of 10
```
(nit) the key thing here is that client A is connected as a `SINGLE_PRIMARY` client, and client B wants to be an `ALL_PRIMARY` client. `ALL_PRIMARY` means that the server (device) itself should accept routes from different clients and select between them, whereas the `SINGLE_PRIMARY` mode means that there is an election ID that is supplied from some external source of elections which designates a single client as the primary, who then injects routes - and the others connected clients do not.

I'd suggest just clarifying here by saying "when client A is connected with `SINGLE_PRIMARY` mode...` replacing the "with election ID" part.

deep-gajjar5 replied:
Acknowledged! Changed the description as per the suggestions.

robshakir reviewed (approved):
@deep-gajjar5 LGTM, thanks for the change. One minor comment about the test description.

robshakir commented:
Thanks!

Today, the test case fails because the fake implementation doesn't support `ALL_PRIMARY` at all - I would suggest just marking the check for the error type a TODO here.

There are also other cases where this is triggered where you might want to check the return values, particularly (from (here)[https://github.com/openconfig/gribigo/blob/main/server/server.go#L467-L481]):

* when a client is connected with FIB ACK and the next one requests RIB ACK (or vice versa)
* when a client is connected requesting the AFT is persisted, and the next one asks for it to not be persisted.

The server should respond to these two with the correct error code.

robshakir commented:
Ah, actually, for this specific case, you can use the [`chk.AllowUnimplemented()`](https://github.com/openconfig/gribigo/blob/88c39e6d854a18e7a44b24c3686ede00bb5b19a4/chk/chk.go#L208) argument.

robshakir commented:
I'm OK with adding the other two cases I discussed in a subsequent change :-)

robshakir commented:
These other two cases are actually covered by (this test)[https://github.com/openconfig/gribigo/blob/88c39e6d854a18e7a44b24c3686ede00bb5b19a4/compliance/election.go#L87] I believe:

https://github.com/openconfig/gribigo/blob/88c39e6d854a18e7a44b24c3686ede00bb5b19a4/compliance/election.go#L87

robshakir commented on compliance/election.go:
```diff
@@ -171,7 +171,11 @@ func TestParamsDifferFromOtherClients(c *fluent.GRIBIClient, t testing.TB, opts
+		t,
+		clientBErr,
+        want, chk.AllowUnimplemented())
+	)
```
Additional `)`.

deep-gajjar5 commented:
> Thanks!
> 
> Today, the test case fails because the fake implementation doesn't support `ALL_PRIMARY` at all - I would suggest just marking the check for the error type a TODO here.
> 
> There are also other cases where this is triggered where you might want to check the return values, particularly (from (here)[https://github.com/openconfig/gribigo/blob/main/server/server.go#L467-L481]):
> 
> * when a client is connected with FIB ACK and the next one requests RIB ACK (or vice versa)
> * when a client is connected requesting the AFT is persisted, and the next one asks for it to not be persisted.
> 
> The server should respond to these two with the correct error code.



> 

Added a

> Thanks!
> 
> Today, the test case fails because the fake implementation doesn't support `ALL_PRIMARY` at all - I would suggest just marking the check for the error type a TODO here.
> 
> There are also other cases where this is triggered where you might want to check the return values, particularly (from (here)[https://github.com/openconfig/gribigo/blob/main/server/server.go#L467-L481]):
> 
> * when a client is connected with FIB ACK and the next one requests RIB ACK (or vice versa)
> * when a client is connected requesting the AFT is persisted, and the next one asks for it to not be persisted.
> 
> The server should respond to these two with the correct error code.

Added a TODO to update the tests once support for ALL_PRIMARY comes in. For now, ignore the current tests since the return error would be UNSUPPORTED_PARAMS instead of PARAMS_DIFFER_FROM_OTHER_CLIENTS.

robshakir commented:
It looks like the tests are still failing here - I'm not sure I understand the last comment, what is the behaviour that wouldn't be supported here? 

If it is that the fake returns `UNSUPPORTED_PARAMS` rather than unimplemented, we could probably fix the fake.

robshakir reviewed (changes requested):
Hi Deep -- let's sync up in real-time on this. I think our wires are a bit crossed with what the right error code to have is -- I think your original change was right, but the issue with `chk.AllowUnimplemented()` that is only fixed in #105 needs to be merged before this would pass.

I'll find a slot for us to chat.

Thanks,
r.

deep-gajjar5 commented:
> Hi Deep -- let's sync up in real-time on this. I think our wires are a bit crossed with what the right error code to have is -- I think your original change was right, but the issue with `chk.AllowUnimplemented()` that is only fixed in #105 needs to be merged before this would pass.
> 
> I'll find a slot for us to chat.
> 
> Thanks, r.

This should be resolved now.

deep-gajjar5 commented:
> Hi Deep -- let's sync up in real-time on this. I think our wires are a bit crossed with what the right error code to have is -- I think your original change was right, but the issue with `chk.AllowUnimplemented()` that is only fixed in #105 needs to be merged before this would pass.
> 
> I'll find a slot for us to chat.
> 
> Thanks, r.

This should be resolved now.

robshakir commented:
LGTM, thanks for the changes.
